apps/src/templates/progress/BubbleFactory.jsx

```diff
@@ -212,6 +212,11 @@ export function getBubbleUrl(
   if (studentId) {
     params.user_id = studentId;
   }
+  if (location.protocol === 'https:') {
+    // The server has given us a http:// URL, but since we are on an https:// connection, we
+    // can avoid a redirect when the bubble is clicked by updating the protocol here.
+    levelUrl = levelUrl.replace('http:', 'https:');
+  }
```

Comment:
I think this url might be coming from the back-end (possibly [here](https://github.com/code-dot-org/code-dot-org/blob/7981d80e8bc7a84e7902658872c5adadc141e32f/dashboard/app/models/script_level.rb#L326)?).  If so, maybe we should change this on the back-end to use the correct protocol or make it relative as Dave suggested?

Reply:
I believe the URL comes from [here](https://github.com/code-dot-org/code-dot-org/blob/7a17ac510c9db89cbc35ac8f657a4b92401b58a3/dashboard/app/helpers/levels_helper.rb#L49).  I am happy to try making this URL protocol-relative, but I was a bit nervous about the impact of such a change.

Reply:
Are we able to make this URL always match the scheme of the original request?

Reply:
Passing the request's protocol from the controller through to the model and then helper seems a bit heavy, but using a protocol-relative URL seems to have worked.

Reply:
Sorry for belaboring this...  I actually didn't know that you could have an url that was just relative to the protocol (and not protocol+host+port) and did some reading on it and at least [some people](https://www.paulirish.com/2010/the-protocol-relative-url/) recommend against this approach.  Given our constrained use, I think it might still be ok but I wanted to ask -- what if we just set the protocol to `https`?  It sounds like we typically enforce this in our production environments (see slack thread in #infrastructure).  Would this break our test/dev environments?

Reply:
> I think what I am missing is why we are using a full url (including protocol and hostname) instead of just using the path directly. do we expect these bubbles to work from a context outside of dashboard server?

if we want to avoid this question, we could also use `CDO.studio_url` as a drop-in replacement for `url_from_path`.

Reply:
actually, `CDO.studio_url` returns a protocol-relative url!
```
[staging] dashboard > CDO.studio_url('foo')
=> "//staging-studio.code.org/foo"
```
that means we are already using protocol-relative urls quite widely in our application. so I would suggest we stick with allowing them in this PR.

I think it would be worth following up to make `url_from_path` use `CDO.studio_url` used instead of `root_url(protocol: ...)`, since `CDO.studio_url` is by far more common in our codebase, and there are only a couple of instances of `root_url` to eliminate.

James, if you want to push to eliminate protocol-relative urls, that could be a bigger change that might be best to take through whatever process we have for dealing with security issues (e.g. posting in #infrastructure).

Reply:
> James, if you want to push to eliminate protocol-relative urls, that could be a bigger change that might be best to take through whatever process we have for dealing with security issues (e.g. posting in #infrastructure)

looking more... given that we already redirect all http requests to https in production, it's hard to imagine how protocol-relative urls could be hurting us. I would expect these only to hurt us if it meant that requesting a page in http allowed an attacker to force that page to load a bunch of assets also in http.